### ISSUE TITLE
sql: bracket subquery results in wrong transaction status

### ISSUE THREAD

**Issue by espadrine** [cockroachdb/cockroach#22208]
**BUG REPORT**

## Reproduction

```sql
CREATE TABLE foo (id INT);
UPDATE foo SET id = 2 WHERE id = (SELECT id FROM [INSERT INTO foo(id) VALUES (1) RETURNING id]);
```

*Expected result:* foo contains a single row with id 2.

*Actual result:* the following error.

```
pq: internal/client/txn.go:885: attempting to use transaction with wrong status or finalized: COMMITTED true
```

*Potentially useful note:* it seems like a bug in [bracket subqueries](https://www.cockroachlabs.com/docs/stable/table-expressions.html#using-the-output-of-other-statements), as the following simpler query has an incorrect result but no errors:

```sql
CREATE TABLE foo (id INT);
SELECT id FROM [INSERT INTO foo(id) VALUES (1) RETURNING id] LIMIT 1;
+----+
| id |
+----+
|  1 |
+----+
SELECT * FROM foo;
+----+
| id |
+----+
+----+
```

## Configuration

```
$ grep -F '[config]' node0/logs/cockroach.log
I180126 15:25:11.916087 1 util/log/clog.go:1031  [config] file created at: 2018/01/26 15:25:11
I180126 15:25:11.916087 1 util/log/clog.go:1031  [config] running on machine: ip-192-168-1-157
I180126 15:25:11.916087 1 util/log/clog.go:1031  [config] binary: CockroachDB CCL v1.1.4 (darwin amd64, built 2018/01/08 18:25:09, go1.9.2)
I180126 15:25:11.916087 1 util/log/clog.go:1031  [config] arguments: [cockroach start --insecure --store node0]
benchmark master
$ grep -F '[config]' node1/logs/cockroach.log
I180129 14:45:35.336207 32031545 util/log/clog.go:928  [config] file created at: 2018/01/29 14:45:35
I180129 14:45:35.336207 32031545 util/log/clog.go:928  [config] running on machine: ip-192-168-1-157
I180129 14:45:35.336207 32031545 util/log/clog.go:928  [config] binary: CockroachDB CCL v1.1.4 (darwin amd64, built 2018/01/08 18:25:09, go1.9.2)
I180129 14:45:35.336207 32031545 util/log/clog.go:928  [config] arguments: [cockroach start --insecure --store node1 --port 26258 --http-port 8081 --join localhost:26257]
benchmark master
$ grep -F '[config]' node2/logs/cockroach.log
I180127 08:10:08.201495 49 util/log/clog.go:928  [config] file created at: 2018/01/27 08:10:08
I180127 08:10:08.201495 49 util/log/clog.go:928  [config] running on machine: ip-192-168-1-157
I180127 08:10:08.201495 49 util/log/clog.go:928  [config] binary: CockroachDB CCL v1.1.4 (darwin amd64, built 2018/01/08 18:25:09, go1.9.2)
I180127 08:10:08.201495 49 util/log/clog.go:928  [config] arguments: [cockroach start --insecure --store node2 --port 26259 --http-port 8082 --join localhost:26257]
```

**Comment by espadrine** [comment]
The main issue here seems related to https://github.com/cockroachdb/cockroach/issues/18806, and is fixed in CCL v2.0-alpha.20180129 @ 2018/01/29 18:31:28 (go1.9).

However, the following odd behavior is still there:

```sql
CREATE TABLE foo (id INT);
SELECT id FROM [INSERT INTO foo(id) VALUES (1) RETURNING id] LIMIT 1;
+----+
| id |
+----+
|  1 |
+----+
SELECT * FROM foo;
-- Actual:
+----+
| id |
+----+
+----+
-- Expected:
+----+
| id |
+----+
|  1 |
+----+
```

Removing `LIMIT 1` works as intended. I am not sure if my expectations are wrong, or if there is still a bug.

**Comment by knz** [comment]
@espadrine thank you for reporting this.

@jordanlewis can you remind me what we decided about insert statements that are not fully consumed?

**Comment by jordanlewis** [comment]
We didn't really finish deciding one way or another. It was contentious. #20919

**Comment by knz** [comment]
Ok we'll discuss this in the coming days and figure out a solution.

**Comment by espadrine** [comment]
Thanks a lot!

Reading through the issues, I am in awe of the complexity you face and the rigor with which you tackle it.

**Comment by knz** [comment]
status update: we discussed and we have a clear picture of how to fix the correctness problem. Will look into implementing on time for v2.0, or at least shortly afterwards.

**Comment by knz** [comment]
will followup in #23156

**Comment by knz** [comment]
@espadrine the original issue you reported was fixed in #20847.

The other issue with the LIMIT is tracked separately already #20732.

I am working on a bug for the latter already; I am going to close this issue since it's now redundant with the other two.

**Comment by espadrine** [comment]
OK, thanks!
